File: 백준/Gold/2636. 치즈/치즈.py
import sys
from collections import deque


def bfs():
    visited = [[False] * m for _ in range(n)]

    q = deque()

    q.append((0, 0))
    visited[0][0] = True
    cnt = 0

    while q:
        x, y = q.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0 <= nx < n and 0 <= ny < m and not visited[nx][ny]:
                if graph[nx][ny] == 0:
                    visited[nx][ny] = True
                    q.append((nx,ny))
                else:
                    graph[nx][ny] = 0
                    cnt += 1
                    visited[nx][ny] = True
    result.append(cnt)
    return cnt

n, m = map(int, sys.stdin.readline().split())

graph = []

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

for _ in range(n):
    graph.append(list(map(int, sys.stdin.readline().split())))

time = 0
result = []

while True:
    if bfs() == 0:
        break
    time += 1

print(time)
print(result[-2])

File: 백준/Gold/2636. 치즈/치즈.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

    static void bfs() {
        boolean[][] visited = new boolean[map.length][map[0].length];

        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{0, 0});
        visited[0][0] = true;
        while (!q.isEmpty()) {
            int[] pos = q.poll();
            int x = pos[0];
            int y = pos[1];

            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];

                if (nx >= 0 && ny >= 0 && nx < map.length && ny < map[0].length && !visited[nx][ny]) {
                    visited[nx][ny] = true;
                    if (map[nx][ny] == 0) {
                        q.add(new int[]{nx, ny});
                    } else {
                        map[nx][ny] = 0;
                        cnt--;
                    }
                }
            }
        }
    }

    static int[][] map;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    static int cnt = 0;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        map = new int[n][m];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                map[i][j] = Integer.parseInt(st.nextToken());
                if (map[i][j] == 1) cnt++;
            }
        }

        int result = 0;
        int time = 0;
        while (true) {
            result = cnt;
            time++;
            bfs();
            if (cnt == 0) break;
        }

        System.out.println(time);
        System.out.println(result);

    }
}
